lib/wit/gitrepo.py

```diff
@@ -183,8 +187,31 @@ def update_dependency(self, dep):
             manifest.write(self.manifest_path())
 
     def checkout(self):
-        proc = self._git_command("checkout", self.revision)
+        current_commit = self.get_commit(self.revision)
+        rev = self.revision
+        if current_commit != self.revision:
+            proc_ref = self._git_command("show-ref")
+            self._git_check(proc_ref)
+            rev_names = proc_ref.stdout.rstrip().split('\n')
+            rev_names = [r.split(' ') for r in rev_names]
+            rev_names = [r[1] for r in rev_names if r[0] == current_commit]
+            rev_names = [r for r in rev_names if not r.startswith('refs/remotes')]
+            rev_names = [verbose_prefix.sub('', r) for r in rev_names]
+
+            suggestions = ''
+            if len(rev_names) > 1:
+                suggestions = ' ({})'.format(', '.join(rev_names))
+
+            if len(rev_names) != 1:
+                rev = self.revision
+                log.info("Checking out '{}' at '{}'{}".format(self.name, rev, suggestions))
+            else:
+                rev = rev_names[0]
+                log.info("Checking out '{}' at '{}' ({})".format(self.name, rev, self.revision))
+
+        proc = self._git_command("checkout", rev)
```

Comment:
I think the act of checking out should also be guarded by `if current_commit != self.revision` right? Otherwise this should still detach HEAD?

Reply:
> the act of checking out should also be guarded by `if current_commit != self.revision` right?

After `GitRepo.clone`, `self.revision == current_commit` even though we aren't checked out. If we added the above check, `GitRepo.clone ; GitRepo.checkout` would never checkout.

Afaik, Git doesn't have a special "not checked out" state; it treats an un-checked-out repo exactly the same as it would if you ran `git checkout ; rm -rf ./*`

> Otherwise this should still detach HEAD?

Only when there are multiple references pointing to the same commit and `HEAD` points to one of those references. I'll fix this (good catch!)

Reply:
Fixed by https://github.com/sifive/wit/pull/106/commits/36acd4a0dc4fb3079f41e014e80a1059bfb7ed03